### DISCUSSION TITLE
Output and Logs Not Captured by Pytest When Using APIRouter

### DISCUSSION THREAD

**Discussion by benjamin-kirkbride** [fastapi/fastapi#6535] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import logging

import uvicorn
from fastapi import APIRouter, FastAPI
from fastapi.testclient import TestClient

import main

logger = logging.getLogger(__name__)


app = FastAPI()
api_router = APIRouter()


@app.get("/")
def root():
    logger.warning("root")


def dependant():
    logger.warning("dependant")


@api_router.get("/")
def router():
    logger.warning("router")
    dependant()


app.include_router(api_router, prefix="/router")

# Test Stuff
client = TestClient(main.app)


def test_root_logging():
    client.get("/")
    assert False


def test_route_logging():
    client.get("/route")
    assert False
```

### Description

When using Pytest, all output (stdout, stderr, and log handlers) are intercepted and captured. By default, upon a test failing, said output is displayed alongside a trackback and other details. This feature greatly assists in resolving the failing test.

stdout and stderr capture appears to not work at all, but I don't care too much about that (though it may be related). What is really problematic though, is that log capturing does not work when the function it is in is called by a `APIRouter` decorated function. I think this has been reported before (https://github.com/tiangolo/fastapi/issues/970), but the original reporter either figured it out and neglected to share the solution or gave up.

To be clear, the issue is not that logs don't work at all. If you run the above snippet, you will see that they do:
![image](https://user-images.githubusercontent.com/3373830/128941004-b30f34b7-a558-4a8e-8261-3b6565aeef7b.png)

The issue is that in Pytest, the logs are not captured and displayed with failed tests in the way they are for functions that are not decorated by `APIRouter`:
![image](https://user-images.githubusercontent.com/3373830/128941218-67b905f0-4dfd-44b8-a769-251f0d1498d7.png)

I'm at somewhat of a loss as to how to proceed. I've tried all kinds of things as far as changing log handlers and such, but none of that seemed to make a difference. Any insight would be greatly appreciated.

### Operating System

Linux

### FastAPI Version

0.68.0

### Python Version

Python 3.9.2

**Comment by adriangb** [comment, marked as answer]
I was unable to reproduce:

<img width="873" alt="image" src="https://user-images.githubusercontent.com/1755071/128946171-7d4a767d-edd4-4163-b6df-3e5972e33260.png">

I had to fix your example (to remove `import main` and other things), but it's basically the same as far as I can tell:

```python
import logging

from fastapi import APIRouter, FastAPI
from fastapi.testclient import TestClient

logger = logging.getLogger(__name__)


app = FastAPI()
api_router = APIRouter()


@app.get("/")
def root():
    logger.warning("root")


def dependant():
    logger.warning("dependant")


@api_router.get("/")
def router():
    logger.warning("router")
    dependant()


app.include_router(api_router, prefix="/router")

# Test Stuff
client = TestClient(app)


def test_root_logging():
    client.get("/")
    assert False


def test_route_logging():
    client.get("/router/")
    assert False
```

**Comment by benjamin-kirkbride** [comment]
Okay.. so your example also works for me, but my original still does not...

**Comment by adriangb** [comment]
Are there any differences I'm missing? As far as I can tell I just removed the invalid import.

**Comment by benjamin-kirkbride** [comment]
:facepalm: 

`client.get("/route")`

Turns out I'm just an idiot.

**Comment by adriangb** [comment]
Happens! Glad we figured it out.

**Comment by benjamin-kirkbride** [comment]
Doesn't explain why I can't get the logs to work in my main app, back to the drawing board I guess.

Thanks!

**Comment by benjamin-kirkbride** [comment]
@adriangb btw, how do you get your pytest looking so slick? :sweat_smile:

**Comment by adriangb** [comment]
https://github.com/Teemu/pytest-sugar

**Comment by adstastic** [comment]
@benjamin-kirkbride did you get your logs to work in the end? Having the same issue.
